test/factory/listing.go

```diff
@@ -0,0 +1,109 @@
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+func NewListing(slug string) *pb.Listing {
+	return &pb.Listing{
+		Slug:               slug,
+		TermsAndConditions: "Sample Terms and Conditions",
+		RefundPolicy:       "Sample Refund policy",
+		Metadata: &pb.Listing_Metadata{
+			Version:            1,
+			AcceptedCurrencies: []string{"tbtc"},
+			PricingCurrency:    "tbtc",
+			Expiry:             &timestamp.Timestamp{Seconds: 2147483647},
+			Format:             pb.Listing_Metadata_FIXED_PRICE,
+			ContractType:       pb.Listing_Metadata_PHYSICAL_GOOD,
+		},
+		Item: &pb.Listing_Item{
+			Skus: []*pb.Listing_Item_Sku{
+				{
+					Surcharge:    0,
+					Quantity:     12,
+					ProductID:    "1",
+					VariantCombo: []uint32{0, 0},
+				},
+				{
+					Surcharge:    0,
+					Quantity:     44,
+					ProductID:    "2",
+					VariantCombo: []uint32{0, 1},
+				},
+			},
+			Title: "Ron Swanson Tshirt",
+			Tags:  []string{"tshirts"},
+			Options: []*pb.Listing_Item_Option{
+				{
+					Name:        "Size",
+					Description: "What size do you want your shirt?",
+					Variants: []*pb.Listing_Item_Option_Variant{
+						{Name: "Small", Image: NewImage()},
+						{Name: "Large", Image: NewImage()},
+					},
+				},
+				{
+					Name:        "Color",
+					Description: "What color do you want your shirt?",
+					Variants: []*pb.Listing_Item_Option_Variant{
+						{Name: "Red", Image: NewImage()},
+						{Name: "Green", Image: NewImage()},
+					},
+				},
+			},
+			Nsfw:           false,
+			Description:    "Example item",
+			Price:          100,
+			ProcessingTime: "3 days",
+			Categories:     []string{"tshirts"},
+			Grams:          14,
+			Condition:      "new",
+			Images:         []*pb.Listing_Item_Image{NewImage(), NewImage()},
+		},
+		Taxes: []*pb.Listing_Tax{
+			{
+				Percentage:  7,
+				TaxShipping: true,
+				TaxType:     "Sales tax",
+				TaxRegions:  []pb.CountryCode{pb.CountryCode_UNITED_STATES},
+			},
+		},
+		ShippingOptions: []*pb.Listing_ShippingOption{
+			{
+				Name:    "usps",
+				Type:    pb.Listing_ShippingOption_FIXED_PRICE,
+				Regions: []pb.CountryCode{pb.CountryCode_ALL},
+				Services: []*pb.Listing_ShippingOption_Service{
+					{
+						Name:              "standard",
+						Price:             20,
+						EstimatedDelivery: "3 days",
+					},
+				},
+			},
+		},
+		Coupons: []*pb.Listing_Coupon{
+			{
+				Title:    "Insider's Discount",
+				Code:     &pb.Listing_Coupon_DiscountCode{"insider"},
+				Discount: &pb.Listing_Coupon_PercentDiscount{5},
+			},
+		},
+	}
+}
+
+func NewCryptoListing(slug string) *pb.Listing {
+	listing := NewListing(slug)
+	listing.Metadata.CoinType = "teth"
+	listing.Metadata.ContractType = pb.Listing_Metadata_CRYPTOCURRENCY
+	listing.Item.Skus = []*pb.Listing_Item_Sku{}
+	listing.Metadata.PricingCurrency = ""
+	listing.ShippingOptions = nil
+	listing.Item.Condition = ""
+	listing.Item.Options = nil
+	listing.Item.Price = 0
+	listing.Coupons = nil
+	return listing
+}
```

Comment:
I prefer instantiating the struct with all fields defined instead of setting them field-wise. Less verbose, too.

Reply:
I do too, but in this case I'm trying to use the base factory listing and override the appropriate fields for crypto listings instead of copying the entire huge pb.Listing definition.

Reply:
Oh, I think I overlooked this being a function call. Saw all the field setting and my mind's eye saw a struct declaration on the first line instead of a method call. Sounds good!